### ISSUE TITLE
Referencing a BOM project in a multi-project build

### ISSUE THREAD

**Issue by mjustin** [spring-gradle-plugins/dependency-management-plugin#164]
I would like to be able to reference a BOM project in a multi-project build in the dependencyManagement section.  As far as I can tell from the documentation, this is currently not supported, correct?

To make things more concrete, assume I have the following project structure:

```
test-parent
  test-bom
  test-module1
  test-module2
```

Assume `test-bom` defines a number of dependencies in its dependencyManagement section and generates a BOM pom using the "maven" or "maven-publish" plugin.  Is there a way to tell `test-module1` & `test-module2` to use the same dependency versions as the BOM project?

In my particular scenario, I'm publishing the generated BOM to my Maven repo, so that it can be used in other projects.  If not for this, I could just define the dependencyManagement section for all projects in test-parent rather than define a separate BOM.

**Comment by wilkinsona** [comment]
You're correct that this isn't currently supported, although you may be able to get it to work by jumping through a couple of hoops. If `test-module1` and `test-module2` were configured such that they triggered the `install` task on `test-bom` before any configurations were resolved, I believe you'd be able to configure them to import the bom from your local Maven repository.

**Comment by scottresnik** [comment]
Based on this ticket, it sounds like dependencies defined in the root project of a multi project build should be available child projects.  Is that the case?  I have a setup similar to the OP, and am defining my dependencies in the "dependencyManagement" block of the root project, yet they are not available to the child projects.  

Putting the dependencyManagement block in the "allprojects" or "subprojects" block does work.  Is this expected?

**Comment by wilkinsona** [comment]
The behaviour you have described is expected. This no inheritance of dependency management from a "parent" project to its child projects.

**Comment by scottresnik** [comment]
Thanks for the reply.
Do you envision parent/root dependency inheritance being part of this enhancement?  If not, having the functionality described by the OP would suffice for my situation.

**Comment by wilkinsona** [comment]
No, I don't. Inheriting configuration between projects isn't really the Gradle way of doing things. Applying the configuring using `allprojects`, `subprojects` or similar, is the recommended way of doing things.

**Comment by wilkinsona** [comment]
I don't have any plans to implement this in the short to medium term. It's something of an edge-case and I haven't seen sufficient demand for it to take priority over other work.

**Comment by eekboom** [comment]
#224 is a duplicate - or at least closely related, isn't it?

**Comment by zxdposter** [comment]
I think it's necessary, such as spring boot micro server, multi project need same version from **spring-cloud-dependencies**, bom cannot be referencing making my project cannot uniform set spring cloud version.

**Comment by mamuf** [comment]
Regarding OP's question, if I understand it correctly, it can be done in at least two ways. Both apply the `test-bom`'s dependency management to your regular modules, but they differ in how all the Maven POMs are generated in the project:

1. Let the Spring Dependency Management plugin do it's magic all the way.
2. Use Gradle's java-platform plugin (since 6.0) and import `test-bom` into the standard module's POM files.

Let's say you have the following project structure (same as OP):

```
test-parent
  test-bom
  test-module1
  test-module2
```

(Note that I'm using what works in Gradle 7 in the examples bellow, haven't tested in older versions.)

test-bom/build.gradle (common for both solutions):
```groovy
apply plugin: 'io.spring.dependency-management'

dependencyManagement {
    // declare your BOM as you would normally
}
```

**Solution 1**

build.gradle (test-parent):
```groovy
plugins {
    id 'java-base'
    id 'io.spring.dependency-management'
}

ext {
    bomProjects = [project(':bom')]
    javaProjects = subprojects - bomProjects
}

configure(javaProjects) {
    apply plugin: 'java-library'
    apply from: "${rootDir}/bom/build.gradle" // apply the dependency management
}
```

So what this does is that the whole `<dependencyManagement>` section from `test-bom` is also included in each of the java subprojects when their POM files are generated. Note that those POMs do not reference the POM of `test-bom` at all.

**Solution 2**

This is almost the same, but note the additional configuration for the `javaProjects`:

build.gradle (test-parent):
```groovy
plugins {
    id 'java-base'
    id 'io.spring.dependency-management'
}

ext {
    bomProjects = [project(':bom')]
    javaProjects = subprojects - bomProjects
}

configure(javaProjects) {
    apply plugin: 'java-library'
    apply from: "${rootDir}/bom/build.gradle" // apply the dependency management

    dependencyManagement {
        generatedPomCustomization {
            // do not add Spring Dependency Management to java module POM files, we use the Gradle platforms (bellow)
            enabled = false
        }
    }
    dependencies {
        // let Gradle Maven Publish plugin add our BOM to each POM's dependencyManagement
        implementation platform(project(':test-bom'))
    }
}
```

So this solution does what you would probably do in Maven. You declare your BOM and then import it via the <dependencyManagement> into each of your submodule's POM.

**But there are some other issues...**

Unfortunately, the issue I have with this plugin, is that when you have a bit more complex setup, it's harder to make it work.

Let's have a project like this:

```
test-parent
  test-bom
  test-dependencies
  test-module1
  test-module2
```

It's almost the same as the original example with the exception that `test-bom` only declares the dependency management of our own project modules. I.e. BOM that would be imported into other people's projects – our project is a library. The `test-dependencies` then declares the dependency management of third party dependencies used by our project, AND the internal modules as well. You can look at those two like the `test-bom` is equivalent to the `spring-framework-bom` and the `test-dependencies` is equivalent to the `spring-boot-dependencies` POM.

Let's see how the gradle files look:

test-bom/build.gradle:
```groovy
dependencies {
    constraints {
        api project(':test-module1')
        api project(':test-module2')
    }
}
```

test-dependencies/gradle.build:
```groovy
apply plugin: 'io.spring.dependency-management'

dependencyManagement {
    imports {
        // coordinates for our test-bom
        mavenBom "test:test-bom:${version}"
        
        // import other BOMs like Spring etc.
        //...
    }
    
    // declare additional managed dependencies as you would normally
    //...
}
```

build.gradle (test-parent):
```groovy
plugins {
    id 'java-base'
    id 'io.spring.dependency-management'
}

ext {
    bomProjects = [project(':bom')]
    javaProjects = subprojects - bomProjects
    version = "1.0.0-SNAPSHOT"
}

configure(javaProjects) {
    apply plugin: 'java-library'
    apply from: "${rootDir}/dependencies/build.gradle" // apply the dependency management
}
```

Now, the problem lies within the `mavenBom "test:test-bom:${version}"` declaration. The problem is the Spring Dependency Management works with the information as if it was an external Maven dependency and ignores the fact it may be an internal module.

Unfortunately, using the `project(':test-bom')` notation is not supported with `mavenBom` (nor is it with the regular `dependency` inside the `dependencyManagement`).

So unless you have already installed the `test-bom` POM file into your (local) Maven repository, the build fails while trying to resolve the dependency.

You can't simply combine the `dependencyManagement` with Gradle's platform constrains like the following example. It will generate bad POM file for Maven publication – with two `<dependencyManagement>` blocks.

```groovy
apply plugin: 'java-platform'
apply plugin: 'io.spring.dependency-management'

dependencies {
    constraints {
        api project(':test-bom')
    }
}

dependencyManagement {
    imports {
        // import other BOMs like Spring etc.
    }
    // declare additional managed dependencies as you would normally
}
```

So I wonder if support for the `project(...)` notation could be implemented...?

**Comment by wilkinsona** [comment]
Thanks for the suggestion, but I don't want to encourage using both the dependency management plugin and Gradle's platform support at the same time. It undermines the long-term goal of Gradle's platform support evolving to be a complete replacement for the dependency management plugin.
